### PR TITLE
Implements martials_like for cross-mod martial arts compatibility

### LINKED ISSUE #45612: Martials_like: A solution to cross-mod martial arts compatibility
# Is your feature request related to a problem? Please describe.
The +1/+2 weapons from Magiclysm only work with vanilla martial arts, when Mythical Martial Arts would pair so well with them(also applies to Aftershock and C.R.I.T. Expansion, but those weapons aren't magic like some of the MythMAs are).

# Describe the solution you'd like
Ability to add martials_like to a weapon's definition, that when applied adds it to every martial art that the weapon it is referencing also works with.

# Describe alternatives you've considered
Using ``copy-from`` to accomplish it.

# Additional context
```
{
    "type": "GENERIC",
    "id": "mace_plus_one",
    "copy-from": "mace",
    "looks_like": "mace",
    "martials_like": "mace",
    "name": "mace +1",
    "proportional": { "price": 3.0, "bashing": 1.1, "cutting": 1.1, "weight": 0.9 },
    "relative": { "to_hit": 1 }
  }
```
-Adds it to Desert Wind, which has the three vanilla maces as options.

### PR DESCRIPTION
**Pull request by Saicchi**
#### Summary
Infrastructure "martials_like for cross mod weapons martial arts compatibility"

#### Purpose of change
Closes #45612
Allows weapons from mod A to be used in martial arts in mod B. It also adds an alternative to extending the list of weapons in a martial arts.

#### Describe the solution
Adds a new field items can have.
`"martials_like": [ "item_id", "item_id2" ]`: Martials arts which have any of the items in the list will also include this item.  
Note that this won't recursively look for items, that means the items in the list must be directly in the martial arts `weapons` entry.

#### Describe alternatives you've considered
Implementing `martials_blacklist`.
`"martials_blacklist": ["mart1_id", "mart2_id"]`: The item won't be added to martial arts  in this list.
This would increase complexity but allows more control. Though I doubt it would be used much due to the way mods are developed.

#### Testing
No issues found.

#### Additional context

I was thinking of adding `martials_like` for all +1 and +2 weapons of Magiclysm, but since its rather big I'll leave it to another PR.

Due to the way `auto_flags_reader` works, `"martials_like": "item_id"` is also valid.

### CODE REVIEW

**Comment by jbytheway** [comment]
An alternative to consider: what if we introduced the idea of a `martial_category`.  Each weapon would be in some categories, and each martial art would list some categories.  If they have any in common then the weapon would be usable.  I think that's more flexible and more robust than this implementation.

File: doc/JSON_INFO.md [hunk 1]
```diff
@@ -2282,6 +2282,7 @@ See also VEHICLE_JSON.md
 "price": 100,                                // Used when bartering with NPCs. For stackable items (ammo, comestibles) this is the price for stack_size charges. Can use string "cent" "USD" or "kUSD".
 "price_postapoc": "1 USD",                       // Same as price but represent value post cataclysm. Can use string "cent" "USD" or "kUSD".
 "material": ["COTTON"],                      // Material types, can be as many as you want.  See materials.json for possible options
+"martials_like": ["weaponid1"]               // (Optional) This weapon can be used in martial arts that use the weapons in the list.
```
Comment (Salty-Panda) [inline doc/JSON_INFO.md:2285]:
```suggestion
"martials_like": ["weaponid1"]               // (Optional) This weapon can be used in martial arts that use the weapons in the list. Not recursive, the weapons specified must be explicitly listed in martial art.
```

File: doc/JSON_INFO.md [hunk 2]
```diff
@@ -2649,6 +2650,7 @@ Any Item can be a container. To add the ability to contain things to an item, yo
 "material": ["iron", "wood"], // Material types.  See materials.json for possible options
 "weight": 907,         // Weight, measured in grams
 "volume": "1500 ml",   // Volume, volume in ml and L can be used - "50 ml" or "2 L"
+"martials_like": ["weaponid1"] // This weapon can be used in martial arts that use the weapons in the list.
```
Comment (Salty-Panda) [inline doc/JSON_INFO.md:2653]:
```suggestion
"martials_like": ["weaponid1"] // This weapon can be used in martial arts that use the weapons in the list.  Not recursive, the weapons specified must be explicitly listed in martial art.
```

File: doc/JSON_INFO.md [hunk 3]
```diff
@@ -2282,6 +2282,7 @@ See also VEHICLE_JSON.md
 "price": 100,                                // Used when bartering with NPCs. For stackable items (ammo, comestibles) this is the price for stack_size charges. Can use string "cent" "USD" or "kUSD".
 "price_postapoc": "1 USD",                       // Same as price but represent value post cataclysm. Can use string "cent" "USD" or "kUSD".
 "material": ["COTTON"],                      // Material types, can be as many as you want.  See materials.json for possible options
+"martials_like": ["weaponid1"]               // (Optional) This weapon can be used in martial arts that use the weapons in the list. Not recursive, the weapons specified must be explicitly listed in martial art.
```
Comment (actual-nh) [inline doc/JSON_INFO.md:2285]:
```suggestion
"price_postapoc": "1 USD",                       // Same as price but represents value post cataclysm. Can use string "cent" "USD" or "kUSD".
"material": ["COTTON"],                      // Material types, can be as many as you want.  See materials.json for possible options.
"martials_like": ["weaponid1"]               // (Optional) This weapon can be used in martial arts that use the (other) weapons in the list. Not recursive; the weapons specified must be explicitly listed by the martial art.
```

File: doc/JSON_INFO.md [hunk 4]
```diff
@@ -2649,6 +2650,7 @@ Any Item can be a container. To add the ability to contain things to an item, yo
 "material": ["iron", "wood"], // Material types.  See materials.json for possible options
 "weight": 907,         // Weight, measured in grams
 "volume": "1500 ml",   // Volume, volume in ml and L can be used - "50 ml" or "2 L"
+"martials_like": ["weaponid1"] // This weapon can be used in martial arts that use the weapons in the list.  Not recursive, the weapons specified must be explicitly listed in martial art.
```
Comment (actual-nh) [inline doc/JSON_INFO.md:2653]:
```suggestion
"martials_like": ["weaponid1"] // This weapon can be used in martial arts that use the (other) weapons in the list.  Not recursive; the weapons specified must be explicitly listed by the martial art.
```

**Comment by Saicchi** [comment]
I've rebase it to cdda-experimental-2021-07-07-1531

**Comment by kevingranade** [comment]
I agree with @jbytheway here, I think referencing another martial art without it being recursive is going to be really confusing, and establishing a single tier of weapon groups that by definition cannot include each other (or that do include each other recursively) is going to work  lot better.

**Comment by actual-nh** [comment]
> I agree with @jbytheway here, I think referencing another martial art without it being recursive is going to be really confusing, and establishing a single tier of weapon groups that by definition cannot include each other (or that do include each other recursively) is going to work lot better.

Some weapons do fall into more than one category, a trench knife being a classic example (blade + brass knuckles), unless I'm misinterpreting what you mean by "weapon groups".

**Comment by jbytheway** [comment]
Weapons can be in multiple groups, and martial arts can reference multiple groups.  That should provide the necessary flexibility, I think.

**Comment by Saicchi** [comment]
So weapons would have a json field like this: `"weapon_category": [ "BLADED", "HALF_LENGTH" ]`, and martial arts would have exactly the same field, and if one ( or all ) the categories in the weapon match one ( or all ) categories in the martial art, then that weapon can be used with the martial art?

**Comment by jbytheway** [comment]
> So weapons would have a json field like this: `"weapon_category": [ "BLADED", "HALF_LENGTH" ]`, and martial arts would have exactly the same field, and if one ( or all ) the categories in the weapon match one ( or all ) categories in the martial art, then that weapon can be used with the martial art?

Yes, that's what I was envisioning.  Specifically, if any category matches any category, then it can be used.

**Comment by Saicchi** [comment]
Should the `weapon_category` replace the `weapons` field in martial arts? Since they will basically do the same thing. They could co-exist, as it wouldn't break any mods that way. Either way I'll close this PR since this deviates considerably from the original solution imo.

**Comment by jbytheway** [comment]
I'd leave support for both.  If you want to start forcing mods to change then you could add a debugmsg that triggers when encountering the old approach, and add a `TEMPORARY until 0.G` comment so we know to go back and remove support after the next stable release.

**Pull request closed without merging**